### ISSUE TITLE
433MHz Transmitter example from the website stops working after repeated execution

### ISSUE THREAD

**Issue by Sirs0ri** [joan2937/pigpio#79]
_tl;dr: Each wave to be sent to a GPIO pin gets an ID which is later treated as byte. This fails if the ID gets bigger than 255._

Hi,
I've been using the example to send 433MHz commands from your website to control a few 433MHz wall outlets via python. In my setup, I have one Raspberry Pi acting as a server that runs some home-automation code and one that has a 433MHz sender and some LEDs attached that are all accessed via pigpio.

This setup has been working fine for a while, but all of a sudden the 433MHz transmitter stopped. Now I'm getting Value Errors every time I try sending a command (my code + the Traceback is below).

After poking around in the code a bit, I noticed that `self._amble = pi.wave_create()` always returns the last value it returned + 3 as new wave-ID (pi.wave_create() returns the previous ID + 1 and is called 2 more times in _make_waves(self)). This ID then ends up in a 'chain' (in send(self, code)), in `pi.wave_chain(chain)` and ultimately in `_pigpio_command_ext(...)` where it's treated as byte (it's the x extending the byte-array ext in `ext.extend(x)`). If the wave-ID is higher than 255 (which has to happen at some point) this will cause a ValueError.

A temporary fix is restarting the Pi pigpiod runs on because that resets the IDs. I'd appreciate if you could look into this and help me find a permanent fix.

Cheers, Maximilian

# My code (python 2.7):
```python
import pigpio

class Transmitter(object):
    """
    A class to transmit the wireless codes sent by 433 MHz
    wireless fobs. This class is taken 1:1 from the '433MHz keyfob RX/TX' 
    example that can be found at 
    http://abyz.co.uk/rpi/pigpio/examples.html#Python%20code
    """

    def __init__(self, pi, gpio, repeats=6, bits=24, gap=9000, t0=300, t1=900):
        """
        Instantiate with the Pi and the GPIO connected to the wireless
        transmitter.

        The number of repeats (default 6) and bits (default 24) may
        be set.

        The pre-/post-amble gap (default 9000 us), short pulse length
        (default 300 us), and long pulse length (default 900 us) may
        be set.
        """
        self.pi = pi
        self.gpio = gpio
        self.repeats = repeats
        self.bits = bits
        self.gap = gap
        self.t0 = t0
        self.t1 = t1

        self._make_waves()

        pi.set_mode(gpio, pigpio.OUTPUT)

    def _make_waves(self):
        """
        Generates the basic waveforms needed to transmit codes.
        """
        wf = []
        wf.append(pigpio.pulse(1 << self.gpio, 0, self.t0))
        wf.append(pigpio.pulse(0, 1 << self.gpio, self.gap))
        self.pi.wave_add_generic(wf)
        self._amble = self.pi.wave_create()

        wf = []
        wf.append(pigpio.pulse(1 << self.gpio, 0, self.t0))
        wf.append(pigpio.pulse(0, 1 << self.gpio, self.t1))
        self.pi.wave_add_generic(wf)
        self._wid0 = self.pi.wave_create()

        wf = []
        wf.append(pigpio.pulse(1 << self.gpio, 0, self.t1))
        wf.append(pigpio.pulse(0, 1 << self.gpio, self.t0))
        self.pi.wave_add_generic(wf)
        self._wid1 = self.pi.wave_create()

    def set_repeats(self, repeats):
        """
        Set the number of code repeats.
        """
        if 1 < repeats < 100:
            self.repeats = repeats

    def set_bits(self, bits):
        """
        Set the number of code bits.
        """
        if 5 < bits < 65:
            self.bits = bits

    def set_timings(self, gap, t0, t1):
        """
        Sets the code gap, short pulse, and long pulse length in us.
        """
        self.gap = gap
        self.t0 = t0
        self.t1 = t1

        self.pi.wave_delete(self._amble)
        self.pi.wave_delete(self._wid0)
        self.pi.wave_delete(self._wid1)

        self._make_waves()

    def send(self, code):
        """
        Transmits the code (using the current settings of repeats,
        bits, gap, short, and long pulse length).
        """
        chain = [self._amble, 255, 0]
        print chain  # added for debugging

        bit = (1 << (self.bits - 1))
        for i in range(self.bits):
            if code & bit:
                chain += [self._wid1]
            else:
                chain += [self._wid0]
            bit = bit >> 1

        chain += [self._amble, 255, 1, self.repeats, 0]
        print chain  # added for debugging

        self.pi.wave_chain(chain)

        while self.pi.wave_tx_busy():
            time.sleep(0.1)

    def cancel(self):
        """
        Cancels the wireless code transmitter.
        """
        self.pi.wave_delete(self._amble)
        self.pi.wave_delete(self._wid0)
        self.pi.wave_delete(self._wid1)

PI = pigpio.pi("192.168.178.56")
TRANSMITTER_PIN = 17  # I have the 433MHz transmitter attached to pin 17.
TRANSMITTER = Transmitter(PI, gpio=TRANSMITTER_PIN)
TRANSMITTER.send(1234)
```

# The Traceback:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 101, in send
  File "C:\Program Files\Python27\lib\site-packages\pigpio.py", line 2226, in wave_chain
    self.sl, _PI_CMD_WVCHA, 0, 0, len(data), [data]))
  File "C:\Program Files\Python27\lib\site-packages\pigpio.py", line 869, in _pigpio_command_ext
    ext.extend(x)
ValueError: byte must be in range(0, 256)
```

**Comment by joan2937** [comment]
Generally pigpio makes no guesses about the users intentions so any tidying up has to be done by the user.

Add a call to TRANSMITTER.cancel() after TRANSMITTER.send(1234).

It's also best if the connection to the daemon is properly terminated which is done by adding a call to PI.stop() as the last line of your script.

p.s. rather than restarting the daemon you could have used the command `pigs wvclr` from a shell to clear all waves.
